### PR TITLE
[vernac] Refactor control attributes and fix bug #10452

### PR DESCRIPTION
**Pull request by ejgallego**
We place control attribute information in its own data structure; this allows to pass the control attributes to the delayed proof functions, fixing #10345 , introduced in #10337

### CODE REVIEW

File: vernac/vernacprop.ml [hunk 1]
```diff
@@ -13,48 +13,27 @@
...
-  )
-
-let rec has_Fail v = v |> CAst.with_val (function
-  | VernacExpr _ -> false
-  | VernacRedirect (_,c)
-  | VernacTime (_,c)
-  | VernacTimeout (_,c) -> has_Fail c
-  | VernacFail _ -> true)
+(* previous semantics of this function were strange, new one is not much better? *)
+let has_Fail { CAst.v } = List.mem ControlFail v.control
```
Comment (SkySkimmer) [inline vernac/vernacprop.ml]: strange in what way?

**Review by SkySkimmer — COMMENTED** [review]
I wonder if it would make sense to have them be parsed as regular attributes (`Vernacexpr.vernac_flags`)

**Comment by ejgallego** [comment]
> I wonder if it would make sense to have them be parsed as regular attributes (`Vernacexpr.vernac_flags`)

I think so, the PR goes towards that direction.

File: vernac/vernacprop.ml (hunk 1 above)
Comment (ejgallego) [reply vernac/vernacprop.ml]: In the sense that `Fail Fail` has a "Fail`, but indeed maybe the comment is confusing, will remove.

File: vernac/g_vernac.mlg [hunk 2]
```diff
@@ -72,16 +72,25 @@ let parse_compat_version = let open Flags in function
     CErrors.user_err ~hdr:"get_compat_version"
       Pp.(str "Unknown compatibility version \"" ++ str s ++ str "\".")
 
+(* XXX: Locations here are messy, fix before merge. *)
```
Comment (gares) [inline vernac/g_vernac.mlg]: so not to forget it

File: vernac/ppvernac.ml [hunk 3]
```diff
@@ -1283,18 +1283,20 @@ let pr_vernac_attributes =
...
-    | VernacRedirect (s, v) ->
-      return (keyword "Redirect" ++ spc() ++ qs s ++ spc() ++ pr_vernac_control v)
-    | VernacTimeout(n,v) ->
-      return (keyword "Timeout " ++ int n ++ spc() ++ pr_vernac_control v)
-    | VernacFail v->
-      return (keyword "Fail" ++ spc() ++ pr_vernac_control v)
+    { control; attrs; expr } ->
+    let pp_control = Pp.mt () in
+    return (pp_control ++ pr_vernac_attributes attrs ++ pr_vernac_expr expr ++ sep_end expr)
+    (* | VernacTime (_,v) ->
```
Comment (gares) [inline vernac/ppvernac.ml]: idem

File: vernac/vernacentries.ml [hunk 4]
```diff
@@ -2719,16 +2721,36 @@ let interp ?(verbosely=true) ~st cmd =
...
   let stack = st.Vernacstate.lemmas in
   let stack = Option.cata (fun stack -> snd @@ Vernacstate.LemmaStack.pop stack) None stack in
+  let () = match pe with
+    | Admitted ->
+      save_lemma_admitted_delayed ~proof ~info
+    | Proved (_,idopt) ->
+      save_lemma_proved_delayed ~proof ~info ~idopt in
+  { st with Vernacstate.lemmas = stack }
+
+let rec interp_qed_delayed_control ~proof ~info ~st ?loc ~control pe : Vernacstate.t =
```
Comment (gares) [inline vernac/vernacentries.ml]: This code seems a to be a duplication of iterp_control above.
It should be factored and abstracted over the base case.

**Review by gares — CHANGES_REQUESTED** [review]
a few reminders + a real request for changes

File: vernac/vernacentries.ml (hunk 4 above)
Comment (ejgallego) [reply vernac/vernacentries.ml]: I tried but that's pretty tricky. I'll give it another go.

File: vernac/ppvernac.ml (hunk 3 above)
Comment (ejgallego) [reply vernac/ppvernac.ml]: Fixed.

File: vernac/vernacentries.ml (hunk 4 above)
Comment (ejgallego) [reply vernac/vernacentries.ml]: Gave it a go, it doesn't look too bad.

**Comment by ejgallego** [comment]
Ready for another round round of reviews, still need to fix some small nits but the overall approach won't change except for review feedback.

File: vernac/vernacentries.ml [hunk 5]
```diff
@@ -2679,22 +2679,29 @@ and vernac_load ~verbosely fname =
...
 
-and interp_control ?proof ~st v = match v with
-  | { v=VernacExpr (atts, cmd) } ->
-    let before_univs = Global.universes () in
-    let pstack = interp_expr ?proof ~atts ~st cmd in
-    if before_univs == Global.universes () then pstack
-    else Option.map (Vernacstate.LemmaStack.map_top_pstate ~f:Proof_global.update_global_env) pstack
-  | { v=VernacFail v } ->
-    with_fail ~st (fun () -> interp_control ?proof ~st v);
+and interp_control_flag ~vernac (f : control_flag) ~st
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml]: This isn't mutually recursive with the rest afaict

File: vernac/g_vernac.mlg (hunk 2 above)
Comment (SkySkimmer) [reply vernac/g_vernac.mlg]: still there

File: vernac/vernacentries.ml (hunk 5 above)
Comment (ejgallego) [reply vernac/vernacentries.ml]: It is not, indeed I forgot to move it upwards [I kept it close for the refactoring as to see both the old and the new].

Moved.

File: vernac/g_vernac.mlg (hunk 2 above)
Comment (ejgallego) [reply vernac/g_vernac.mlg]: Yup, will improve once the rest of code (that is, use points) is known to be stable.

File: stm/stm.ml [hunk 6]
```diff
@@ -1979,13 +1979,15 @@ end = struct (* {{{ *)
...
-        | VernacRedirect (_,e) -> find ~time ~batch ~fail e
-        | VernacFail e -> find ~time ~batch ~fail:true e
-        | e -> CAst.make ?loc e, time, batch, fail) v in
-      find ~time:false ~batch:false ~fail:false e in
+    let cl, time, batch, fail =
+      let rec find ~time ~batch ~fail cl = match cl with
+        | ControlTime batch :: cl -> find ~time:true ~batch ~fail cl
+        | ControlRedirect _ :: cl -> find ~time ~batch ~fail cl
+        | ControlFail :: cl -> find ~time ~batch ~fail:true cl
+        | ControlTimeout _ :: cl
```
Comment (gares) [inline stm/stm.ml]: Why are you stripping Timeout? The old code was leaving it these, so that the worker would execute it.

**Review by gares — CHANGES_REQUESTED** [review]
Timeout changed meaning

File: stm/stm.ml (hunk 6 above)
Comment (ejgallego) [reply stm/stm.ml]: Let me see, likely a programming error on my part.

Comment (gares) [reply stm/stm.ml]: The old code is brittle, and the possibility to nest controls as you like makes it very hard to give them a sound meaning. But the new code is just different, and I'm pretty sure I did put in the test suite a `par: Timeout x tac.`

Comment (ejgallego) [reply stm/stm.ml]: Should be fixed now.

Comment (ejgallego) [reply stm/stm.ml]: > I'm pretty sure I did put in the test suite a `par: Timeout x tac.` or `Timeout par: tac` (whatever parses ;-) )

The test-suite passed both ways [dropping the timeout and without], so I dunno but this particular bit didn't have any effect on it.

**Comment by ejgallego** [comment]
Ok, I think this is ready for an assignee, all the comments have been addressed TTBOMK and I think the code looks in reasonable shape. We can improve more of course, and eventually merge with attributes.

Note that I think I introduced a bug which basically reversed control operators so `Fail Timeout` was parsed as `Timeout Fail` and indeed the test-suite didn't catch those except for windows, so we may want to review what is going on with those tests.

**Comment by ejgallego** [comment]
Ping folks.

**Comment by maximedenes** [comment]
> I wonder if it would make sense to have them be parsed as regular attributes

I thought attributes were meant to not be semantically ordered.

**Comment by ejgallego** [comment]
> I thought attributes were meant to not be semantically ordered.

That's a good question! Order seems to be tricky, for example in OCaml there are a few ppx where order does matter [for example you need to run ppx_import before you run ppx_derive].

File: stm/stm.ml (hunk 6 above)
Comment (gares) [reply stm/stm.ml]: My bad, there are tests with Time, not with Timeout.

**Review by gares — APPROVED** [review]
stm ok

**Comment by ejgallego** [comment]
Still looking for an assignee, anyone from the reviewers list should be good.

File: vernac/vernacentries.ml [hunk 7]
```diff
@@ -2253,7 +2253,30 @@ let locate_if_not_already ?loc (e, info) =
...
+    (fn : st:Vernacstate.t -> Vernacstate.LemmaStack.t option) =
+  match f with
+  | ControlFail ->
+    with_fail ~st (fun () -> fn ~st);
+    st.Vernacstate.lemmas
+  | ControlTimeout timeout ->
+    vernac_timeout ~timeout (fun () -> fn ~st) ()
+  | ControlTime batch ->
+    let header = if batch then pr_time_header vernac else Pp.mt () in
+    System.with_time ~batch ~header (fun () -> fn ~st) ()
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml]: I don't like having to pass the vernac just for printing, it makes the type of interp_control_flag look misleading. Maybe we could pass the pre-printed header instead? (probably wrapped in lazy)

Comment (SkySkimmer) [reply vernac/vernacentries.ml]: Also maybe we should just print the expr without the attributes/controls. That change may be outside the scope of this PR though.

File: vernac/vernacentries.ml [hunk 8]
```diff
@@ -2717,16 +2735,30 @@ let interp ?(verbosely=true) ~st cmd =
...
+
+let interp_qed_delayed_control ~proof ~info ~st ?loc ~control pe : Vernacstate.t =
+  let vernac = CAst.make ?loc { control; attrs = []; expr = VernacEndProof pe } in
+  let stack =
+    List.fold_right (fun flag fn -> interp_control_flag ~vernac flag fn)
+      control
+      (fun ~st -> interp_qed_delayed ~proof ~info ~st pe)
+      ~st in
+  Vernacstate.Proof_global.set stack [@ocaml.warning "-3"];
+  Vernacstate.freeze_interp_state ~marshallable:false
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml]: These last 2 lines didn't use to be there AFAICT, why the change?

File: vernac/vernacentries.ml [hunk 9]
```diff
@@ -2717,16 +2735,30 @@ let interp ?(verbosely=true) ~st cmd =
...
+      save_lemma_proved_delayed ~proof ~info ~idopt in
+  stack
+
+let interp_qed_delayed_control ~proof ~info ~st ?loc ~control pe : Vernacstate.t =
+  let vernac = CAst.make ?loc { control; attrs = []; expr = VernacEndProof pe } in
+  let stack =
+    List.fold_right (fun flag fn -> interp_control_flag ~vernac flag fn)
+      control
+      (fun ~st -> interp_qed_delayed ~proof ~info ~st pe)
+      ~st in
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml]: Is it just me who likes to have the `in` on its own line for multiline letins?

**Review by SkySkimmer — APPROVED** [review]
Seems reasonable overall.

File: vernac/ppvernac.ml [hunk 10]
```diff
@@ -1276,24 +1276,27 @@ let rec pr_vernac_flag (k, v) =
...
 
+let pr_control_flag (p : control_flag) =
+  let w = match p with
+    | ControlTime _ -> keyword "Time"
+    | ControlRedirect s -> keyword "Redirect" ++ spc() ++ qs s
+    | ControlTimeout n -> keyword "Timeout " ++ int n
+    | ControlFail -> keyword "Fail" in
+  w ++ spc ()
+
+let pr_control flags = Pp.prlist pr_control_flag flags
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply vernac/ppvernac.ml]: Fixed.

File: vernac/ppvernac.ml [hunk 11]
```diff
@@ -1276,24 +1276,27 @@ let rec pr_vernac_flag (k, v) =
...
-      try pr_vernac_control v
-      with e -> CErrors.print e
+let pr_vernac_control v =
+  let return = tag_vernac v in
+  match v.v with
+    { control; attrs; expr } ->
+    return (pr_control control ++ pr_vernac_attributes attrs ++ pr_vernac_expr expr ++ sep_end expr)
+
+let pr_vernac v =
+  try pr_vernac_control v
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply vernac/ppvernac.ml]: Fixed, thanks.

File: vernac/vernacentries.ml (hunk 7 above)
Comment (ejgallego) [reply vernac/vernacentries.ml]: Indeed I'd be afraid of the change as some users do rely on the printing to be exact, I dunno. I suggest we rename the argument for now?

File: vernac/vernacentries.ml (hunk 8 above)
Comment (ejgallego) [reply vernac/vernacentries.ml]: I think this was a bug in the previous implementation. In fact even the current one seems wrong, right? An `unfreeze` seems to be missing.

Maybe we should refactor the code so interp is shared.

The thing is that I'm quite close to remove the `interp_delayed` stuff in a separate branch reworking proof closing, but meanwhile we should try to make this sound.

File: vernac/vernacentries.ml (hunk 9 above)
Comment (ejgallego) [reply vernac/vernacentries.ml]: Fixed, but indeed achieving consistency on this stuff is hard, we should switch to `ocamlformat` at some point.

File: vernac/vernacentries.ml (hunk 8 above)
Comment (ejgallego) [reply vernac/vernacentries.ml]: I think this was a bug in the previous implementation. In fact even the current one seems wrong, right? An `unfreeze` seems to be missing.

Maybe we should refactor the code so interp is shared.

The thing is that I'm quite close to remove the `interp_delayed` stuff in a separate branch reworking proof closing, but meanwhile we should try to make this sound.

Comment (ejgallego) [reply vernac/vernacentries.ml]: Ok, I've added a commit that tries to address this in a more systematic way.
